### ISSUE TITLE
Pasting emoji into string crashes Visual Studio 2019

### ISSUE THREAD

**Issue by watfordjc** [EWSoftware/VSSpellChecker#218]
Yesterday I was trying to paste an emoji into a text string (to overwrite the img alt Twitter uses when copying text from a Tweet), and every time I pasted the fire engine emoji Visual Studio crashed. After the file recovery process, a banner at the top of the window suggested disabling this extension.

Extension enabled: Visual Studio crashes if I perform the action
Extension disabled: Visual Studio doesn't crash if I perform the action

## Code Sample

This is the [relevant line of my code](https://github.com/watfordjc/csharp-text-formatter/blob/d8afb34415f515dc355fe14bcac3e3f0bcd1b20b/TextFormatter/MainWindow.xaml.cs#L189), although I'm fairly sure the only thing of relevance is the string value:

```csharp
Interop.UnsafeNativeMethods.DrawTextFromString(canvas, "Blue watch are getting ready to handover to green watch for the night. We will be back at 20:00 tomorrow night. Stay safe 🚒😀🚒", centredTopLeft.X, (int)Math.Round(startY), canvasDimensions.Width - (centredTopLeft.X * 2), canvasDimensions.Height - (int)Math.Round(startY), false, "Segoe UI Emoji", 90.0f, 700, "en-GB", brushes["textBrush"]);
```

This is the string copied from Twitter [[Link to Tweet](https://twitter.com/HertsFRSControl/status/1299414833910218754) to comply with TOS] surrounded with double quotes:

```csharp
"Blue watch are getting ready to handover to green watch for the night. We will be back at 20:00 tomorrow night. Stay safe Fire engineGrinning faceFire engine"
```

### Thoughts

I'm not sure if just pasting 🚒 into a string causes the issue, whether pasting 🚒 over the selected text "Fire engine" causes the issue, or if pasting any emoji causes the issue - I haven't extensively tested it.

I'm also not sure at what point in the pasting/spellchecking process a change in text encoding is involved, but I'm fairly sure the emoji I copied was in UTF-8 (it was from a Web page) and ```string``` is UTF-16.

I know with the extension disabled the monochrome emoji is the same colour as string text when pasted (red\*) and turns purple\* after some delay, so there is also syntax highlighting involved.

\* Visual Experience Color Theme: Blue; Editor Color Scheme for C#: Visual Studio 2017

## Software

* Windows 10 Version 2004, Build 19041.450 (Microsoft Support had me do an in-place upgrade 4 days ago)
* Visual Studio Community 2019 version 16.7.1 with .NET Framework 4.8.04084
* Visual Studio Spell Checker (VS2017 and later) version 2020.8.24.0

**Comment by EWSoftware** [comment]
Someone reported the same issue pasting a trash can Unicode character into a string (#216).  As with that case, I'm unable to duplicate this in either VS 2017 or VS 2019.  The version of Visual Studio and theme don't make a difference as far as I can tell.  What language dictionary are you using? It's possible configuration settings could make a difference. Do you recall if you changed any of the configuration settings?

One thing you can try to work around the problem is to go into the configuration settings in the [General category](http://ewsoftware.github.io/VSSpellChecker/html/b4a8726f-5bee-48a4-81a9-00b1be332607.htm) and set the Ignored Character Class option to either ignore non-Latin or ignore non-ASCII characters.  That should cause it to ignore the emoji characters.

**Comment by watfordjc** [comment]
Ignore non-Latin stops the crashes.

---

## Configuration Settings

Global Dictionary: en-GB
General Settings: everything is Yes except Treat underscores as separators. Ignored character class: include all words.
C# Options: everything is No
Excluded expression: ```\[JsonPropertyName\(".*?"\)]```
Code Analysis Dictionaries: All Yes except casing exceptions. Recognised word handling: Treat all as ignored words.

### Research

* Adding the specific emoji to excluded expressions and restarting VS seems to stop the crashing (i.e. it didn't crash, but I haven't worked out how to make crashing a certainty).
* Reproducibility isn't as simple as I thought.
  * If pasting into a string doesn't cause a crash, switching tab and back and pasting into the same string might cause a crash.
  * If pasting into a string doesn't cause a crash, pasting into a different string on the next line might cause a crash.
  * So far, all the strings I have been pasting into that cause crashes are string variables or properties.
  * Very limited sample size, but files under source control appear to crash when pasting under fewer circumstances.
  * Undo/redo buffer isn't a factor and it doesn't look like time is either. Only considered because a file under source control wouldn't crash on first paste, but crashes can occur on first paste after restarting VS with an empty undo buffer so irrelevant.
  * Given pasting into the same string after every crash can cause another crash reliably (until it doesn't), and sometimes it takes a period of tab-switching and/or looking for a string to paste into that will cause a crash, it feels like it is a race condition involving some kind of state tracking. If it doesn't crash on first try, it will eventually under apparently the same conditions or it will appear to (frequency illusion/recency bias).
* I am wondering if adding the (emoji) UTF-8 code point range or the UTF-16 surrogate pair range to excluded expressions would rule out character encoding changes during pasting or narrow down the timing of the issue.

## Potentially Relevant (also analysing what I type, as I type)

### Other extensions

* .NET Portability Analyzer
* Code Alignment
* Visual Studio IntelliCode

### Project dependencies

* Analyzers
  * Humanizer
  * Microsoft.CodeAnalysis.VersionCheckAnalyzer
  * Microsoft.CodeQuality.Analyzers
  * Microsoft.CodeQuality.CSharp.Analyzers
  * Microsoft.NetCore.Analyzers
  * Microsoft.NetCore.CSharp.Analyzers
  * Microsoft.NetFramework.Analyzers
  * Microsoft.NetFramework.CSharp.Analyzers
* Packages
  * Microsoft.CodeAnalysis.FxCopAnalyzers

### Project Settings

* Code Analysis
  * Run on build
  * Run on live analysis

## Probably Irrelevant

* [*.sln Solution File](https://github.com/watfordjc/csharp-stream-controller/blob/master/StreamController.sln)
* [*.csproj Project File](https://github.com/watfordjc/csharp-stream-controller/blob/master/StreamController/StreamController.csproj)
* [Classifications.config](https://gist.github.com/watfordjc/6dc47a807929de2d0d8521b52efbf95e)

**Comment by NoelAbrahams** [comment]
Ran into this today. The act of pasting did not crash VS, but when the adjoining text was edited, attempting to save the file caused the crash. The text is in a `.json` file and the emoji was  😄 (that's the emoji - not me smiling. This is me smiling 😄 )

**Comment by EWSoftware** [comment]
I was able to do some more testing on this and was finally able to duplicate the issue by using the English UK dictionary rather than the English US dictionary.  Part of the problem is that the emojis are being included as part of the preceding or following word.  The main failure occurs because NHunspell crashes when getting suggestions for a word containing an emoji character when it's using the UK dictionary.

The solution will be something along the lines of updating the word splitter to check for surrogate pairs in the string being split and if the pair is within the known ranges of emoji characters, split the word there rather than including the emoji character as part of the word.

A workaround for now is to set the [Ignored Character Class option](http://ewsoftware.github.io/VSSpellChecker/html/b4a8726f-5bee-48a4-81a9-00b1be332607.htm#IgnoredCharClass) to ignore non-Latin or non-ASCII characters.  A side effect is that it may miss misspelled words that abut emojis until the fix above is implemented.

**Comment by EWSoftware** [comment]
This issue has been fixed and an new release is available (v2021.1.23.0).
